cirq/optimizers/cphase_to_fsim.py

```diff
@@ -0,0 +1,224 @@
+
+    sin_half_phi = np.sin(phi / 2)
+    cos_half_phi = np.cos(phi / 2)
+    sin_theta = np.sin(theta)
+    cos_theta = np.cos(theta)
+
+    #
+    # Step 1: find alpha
+    #
+    denominator = (sin_theta - sin_half_phi) * (sin_theta + sin_half_phi)
+    if abs(denominator) < atol:
+        raise ValueError(
+            f'{fsim_gate} cannot be used to decompose CZPowGate because '
+            'sin(theta)**2 is too close to sin(phi/2)**2 '
+            f'(difference is {denominator}).')
+
+    # Parametrization of CZPowGate by a real angle is a non-injective function
+    # with the preimage of cphase_gate infinite. However, it is sufficient to
+    # check just two of the angles against the constraints of the decomposition.
+    canonical_delta = -np.pi * (cphase_gate.exponent % 2)
+    for delta in (canonical_delta, canonical_delta + 2 * np.pi):
+        sin_quarter_delta = np.sin(delta / 4)
+        numerator = (sin_quarter_delta - sin_half_phi) * (sin_quarter_delta +
+                                                          sin_half_phi)
+        sin_alpha_squared = numerator / denominator
+        if 0 <= sin_alpha_squared <= 1:
+            break
+    else:
+        rs = compute_cphase_exponents_for_fsim_decomposition(fsim_gate)
+        raise ValueError(
+            f'{cphase_gate} cannot be decomposed into two {fsim_gate}. '
+            f'Valid intervals for canonical exponent of CZPowGate: {rs}.')
+    assert 0 <= sin_alpha_squared <= 1
+    alpha = np.arcsin(np.sqrt(sin_alpha_squared))
+
+    #
+    # Step 2: find xi and eta
+    #
+    tan_alpha = np.tan(alpha)
+    xi = np.arctan2(tan_alpha * cos_theta, cos_half_phi)
+    eta = np.arctan2(tan_alpha * sin_theta, sin_half_phi)
+    if delta < 0:
+        eta += np.pi
+
+    #
+    # Step 3: synthesize output circuit
+    #
+    circuit = circuits.Circuit(
+        # Local X rotations to convert Γ1⊗I − iZ⊗Γ2 into exp(-i Z⊗Z δ/4)
+        ops.rx(xi).on(q0),
+        ops.rx(eta).on(q1),
+
+        # Y(θ, φ) := exp(-i X⊗X θ/2) exp(-i Y⊗Y θ/2) exp(-i Z⊗Z φ/4)
+        fsim_gate.on(q0, q1),
+        ops.rz(phi / 2).on(q0),
+        ops.rz(phi / 2).on(q1),
+        ops.GlobalPhaseOperation(np.exp(1j * phi / 4)),
+
+        # exp(i X1 α)
+        ops.rx(-2 * alpha).on(q0),
+
+        # Y(-θ, φ) := exp(i X⊗X θ/2) exp(i Y⊗Y θ/2) exp(-i Z⊗Z φ/4)
+        ops.Z(q0),
+        fsim_gate.on(q0, q1),
+        ops.rz(phi / 2).on(q0),
+        ops.rz(phi / 2).on(q1),
+        ops.GlobalPhaseOperation(np.exp(1j * phi / 4)),
+        ops.Z(q0),
+
+        # Local X rotations to convert Γ1⊗I − iZ⊗Γ2 into exp(-i Z⊗Z δ/4)
+        ops.rx(-eta).on(q1),
+        ops.rx(xi).on(q0),
+
+        # Local Z rotations to convert exp(-i Z⊗Z δ/4) into desired CPhase.
+        ops.rz(-delta / 2).on(q0),
+        ops.rz(-delta / 2).on(q1),
+        ops.GlobalPhaseOperation(np.exp(-1j * delta / 4)),
+    )
+
+    MergeSingleQubitGates().optimize_circuit(circuit)
+    DropEmptyMoments().optimize_circuit(circuit)
```

Comment:
Could you skip those, and let user decide how to optimize circuit?

Reply:
I did this originally for consistency with [another optimizer](https://github.com/quantumlib/Cirq/blob/master/cirq/optimizers/two_qubit_to_fsim.py), but on second thought I agree we should leave this up to the user.

(I also considered adding a boolean flag, but calling an additional optimizer is arguably so easy that this just pollutes the signature for little benefit.)

I'll fix the other optimizer for consistency in follow-up PRs (they are out-of-scope here).

Reply:
I promised a PR, but filed #3050 instead in case anyone disagrees or relies on current behavior.